### PR TITLE
Initial set of json data

### PR DESCRIPTION
**Pull request by adgaudio**
Hi @jkeen !

I ported all the couriers we had, as well as added support for recognizing the S10 countries.  I hooked up my java library [MysteryTrackingNumber](https://github.com/adgaudio/MysteryTrackingNumber) to use this json data as a submodule, and my tests are passing.  It was a lot off work!  I also have 2d barcode support in the pipeline.
- **gen_s10_countries.py** and **lint_json.sh** are helpful scripts for maintaining this repo.  We could add travis tests to check linting if necessary.

Could you please take a look at the README and files, and let me know what you think?  I tried to make it cross-language compatible, so curious how this works out for you.  You may have some minor hang up on the string formatting and regex named groups, but I don't know of suitable cross-language standards.

Alex

### CODE REVIEW

**Comment by jkeen** [comment]
Hey Alex! I like this! There's a few things I think we can improve upon this even more before saying this is _the_ way going forward. Some of this feels complex (some of which is unavoidable) but I think we can make it simpler.

- I love the idea of having named `regex` groups (!), but the `P` in the groups like `?P<ApplicationIdentifier>` throws an error in ruby, but works fine without the `P`. What's that doing for you?

- I'm still noodling on how best to format these regexs when it comes to searching for them in a big block of text and validating an individual mystery tracking number. In the original code I had two different regexes, where the search pattern ignored all spaces since sometimes these numbers are spaced out and sometimes they aren't.

```ruby
      VERIFY_PATTERN = /^([A-Z]{2}[0-9]{9}[A-Z]{2})$/
      SEARCH_PATTERN = [/(\b([A-Z]\s*){2}([0-9]\s*){9}([A-Z]\s*)([A-Z]\s*)\b)/]
```

I was trying to avoid that, but it might be unavoidable

- I think we need a non-language specific read me file outlining some check digit algorithms for each tracking number described in plain english, kinda like those specs do. This repo should become _the place_ to go when needing to decode a tracking number, with those externally linked docs only as reference/backup. (I put links in my tracking number repo long ago that are now broken because USPS doesn't know how to 301)

- Why do we need the `serial_number_parser` field? 

- I'd also much rather have the repetition for clarity and ease of reading in the file over the `_inherits` thing. On that same note the "test numbers" and additional information for a courier I think could be integrated right into the same block of json in `couriers.json`. Seeing some good example numbers next to the data about the number feels like it could helpful, and having all the information in one place would be nice for clarity.

**Comment by jkeen** [comment]
Here's an abbreviated example of what I was thinking for the single block of json with all the info in it. And having the named groups is a neat way to match the string with a way to pull out additional information from it: 

```json
{
  "name": "S10",
  "tracking_url": null,
  "regex": "(<application_identifier>[A-Z]{2})(?<serial_number>[0-9]{8})(<check_digit>[0-9])(<country_code>[A-Z]{2})",
  "validation": {
    "checksum": {
      "name": "s10"
    },
    "country_code": "exists"
  },
  "additional": {
    "application_identifier": [
      {
        "name": "Letter Post Express",
        "regex": "L[A-Z]",
        "description": ""
      },
      {
        "name": "Letter Post M-bag",
        "regex": "M[A-Z]",
        "description": "Direct sacks of printed matter sent to a single foreign addressee at a single address"
      },
      {
        "name": "Letter Post IBRS",
        "regex": "Q[A-M]",
        "description": "International Business Reply Service"
      }
    ],
    "country_code": {
      "AF": {
        "country": "Afghanistan",
        "courier": "Afghan Post",
        "courier_url": "http://postalcode.afghanpost.gov.af/",
        "upu_reference_url": "http://www.upu.int/en/the-upu/member-countries/southern-asia-and-oceania/afghanistan.html",
      },
      "AL": {
        "country": "Albania",
        "courier": "Posta Shqiptare",
        "courier_url": "http://www.postashqiptare.al/",
        "upu_reference_url": "http://www.upu.int/en/the-upu/member-countries/western-europe/albania.html",
      },
      "DZ": {        
        "country": "Algeria",
        "courier": "Algérie Poste",
        "courier_url": "http://www.poste.dz/codepostal/",
        "upu_reference_url": "http://www.upu.int/en/the-upu/member-countries/africa/algeria.html",
      } 
    }
  },
  "test_numbers": {
    "valid": [
      "RB123456785US",
      "RB123456785GB",
      "RB123456785US",
      "RB123456785CV",
      "RB123456785CF"
    ],
    "invalid": [
      "RB123456786US"
    ]    
  }
}
```

**Comment by jkeen** [comment]
⬆️ 

We could just define that the "additional" block can contain the named groups in the regex matching, and then can define additional information about those groups. 

If the value at the key is an array, then each item in the array should have a "regex" key to match against. If the value at the key is a hash, then the value of the named group should be represented within. 

This way for the S10 spec we can define in the "validation" key that in order for the number to be valid, it needs to pass the checksum and the country has to exist. Does that make sense?

**Comment by adgaudio** [comment]
Wonderful!

#### Items you requested/suggested that I will implement:

1. Convert (?P<named_group>.*)  to (?<named_group>.*) 
  - the 'P' was a Python named group.  Java doesn't actually support the P, either.
2.  _inherits  -- remove it from couriers.json.  (further discussion of json file structure below)
3. test_couriers.json:  remove the file and put its data in couriers.json, as you specified.
4. "additional" field - I love this idea.  It would get rid of s10_service_indicator.json, which is great.   I imagine "additional" can be used to process any extra named groups in the regular expression.

#### Items that we should talk more about:

1. The search pattern to handle whitespaces.  As far as I know, a whitespace will never be explicitly coded into a tracking number.  Therefore, could you simply strip all whitespaces from the string before parsing it?  What do you think of:

        re.search(  '([A-Z]{2}[0-9]{9}[A-Z]{2})',   MYSTRING.replace(' ', '')  )

2. Reference check digit algorithms:  Are you suggesting we define a pseudo-code implementation in this repo?  I almost did that, but then hesitated because the implementation wouldn't be tested or necessarily remain up to date.  If that is what you are suggesting, let's do it!  To tell you the truth, even with the documentation, I would still use the implementation in a library as my reference, but I see the benefit.  Let me know more specifically what you are thinking we should do here.

3. `serial_number_parser` is kind of a hack I made up to support different ways of converting the serial number string (from regex) into an array of integers.  The idea is that for FedEx SmartPost and UPS the serial number needs to be manipulated before being passed to the check digit algorithm.  I didn't want to create two distinct check digit methods just for smartpost and ups, and I didn't want to make the check digit algorithms perform string parsing, so I created serial_number_parser ...  got any other ideas?

4. s10_countries.json: your example suggests merging s10_countries.json into couriers.json. 
 I'm leaning towards keeping it separate.  The countries list is automatically generated, and we don't want people to think they can modify that list.  It's also very long, making the file difficult to read.  On the other hand, if we do want to merge the country data into couriers.json, we can add a frivolous field like `{"_comment": "autogenerated. do not modify this object", ...}` in every country block.  That doesn't help with the length though.  Thoughts?

5. s10_overrides.json - let's get rid of this by having the script that generates s10_countries.json read it and embed its values into the generated output.  Do you agree?

**Comment by adgaudio** [comment]
6. Another talking point: maybe couriers.json can become a directory, where each file in that directory contains one or more couriers.  Then, our libraries can read the files in json data in that directory and concatenate them together into one large json object.  Thoughts?

**Comment by adgaudio** [comment]
:arrow_up: tests pass with these changes in MysteryTrackingNumber.  I will wait on your comments for further changes.  (Note: I don't actually use service indicator yet)

**Comment by jkeen** [comment]
1. Stripping the whitespace out of the search string kinda blew my mind, initially. That is a great simple solution for most cases but I want to make sure I don't break any existing functionality with the `tracking_number` gem and the one case that I think that would fail at is if multiple tracking numbers are listed in a row. Detecting whitespace before and after the string pretty crucial in that case.

2. Yes! For our own reference and for future implementers. So we'd list them out with the key we specify like "s10_mod10" or whatever and then a description of what needs to be done to calculate it. 

3. I was leaning towards having the validation take care of that. Another option to pass into one of those validators—`prepend:` or something? That is a really annoying thing with that one tracking number. I wonder if it needs to split up into two different types of tracking numbers? 

4, 5, 6.  Interesting idea re: couriers becoming a folder, split up by each courier. That could be good. Then s10.json could be long and include all that generated country data. I'm not deterred by having it be a lengthy file, as long as it's clear. I think having it together makes sense, though, since it wouldn't require the implementer to do anything particularly different for that one case and could use the same `additional` strategy as other attributes. I don't even think the comment that it's autogenerated matters, since we might have to make corrections to individual countries as we go on if we find out better information (e.g. Royal Mail for GB instead of whatever that file originally said)

**Comment by adgaudio** [comment]
1. What about trying stripping whitespace and applying a regex string search rather than a match (search does not imply \^ or \$).  That way, the barcodes go character by character through the string searching for barcodes.  I'm not sure how this will impact speed, but might be worth a try.    For a speed-up, a second thing you might do is wrap the regex in "\b" identifiers, though two tracking numbers separated by a space would not be recognized.

        re.search(
            "\b%s\b" % "pattern",
            "string with pattern in it... pattern again".replace(" ", ""))

2. I will write the reference documentation as discussed

3. serial_number_parser:  I vote keep it for now...  Unfortunately, we can't just use a regex or validator because the prepend logic is post-processing that happens after regex match but before check digit calc.  "USPS91 (20)" explicitly needs to prepend a number.  UPS has a particular processing method (that I applied by default to everyone).   And I didn't realize that OnTrac also needs the prepend logic.  Therefore, I vote keep the serial_number_parser for now...

4-6.  I will break up couriers.json into separate files under a directory called couriers/.   
 For S10 countries, I will merge the countries with their overridden values as one flat list and put it in the additional field of couriers/s10.json.  I won't keep track of whether we did things like rename "United States..." with "USPS (S10)".

**Comment by adgaudio** [comment]
2 and "4-6" done.  If this all looks good to you, then your turn to implement into your ruby library.

I'll add that in I made some improvements to the tracking numbers so we now capture 34 digit usps numbers (and as a result fedex smartpost no longer exists), and I also added additional support for fedex express (I think there is one more tracking fedex express tracking number we are missing).

**Comment by jkeen** [comment]
Thanks for doing all the hard work on this! I've been consumed with other things and appreciate you picking up the slack. 

1. Yeah, still some stuff to be ironed out there. I'll have to see how it shakes out when I integrate this into `tracking_number`. I had a proof of concept going and wrapping each character group in `\b`'s wasn't doing the trick as I wanted it to. 

4-6. I think if we're going to have separate files for each carrier, we shouldn't have `misc.json` for UPS and ontrac. Those should be `ups.json` and `ontrac.json` for consistency, even though they're small.  

3. 90% of the time "for now" turns into "forever" in my experience. 😛  BUT I don't come to the table empty handed. Since `serial_number_parser` is only gonna be used to validate, let's combine it with the `validation` key like I suggested in that example json. Then it all fits together in the same logical group, like so:

So for USPS91 it'd be:
```
"validation": {
    "checksum": {
      "name": "mod10",
      "evens_multiplier": 3,
      "odds_multiplier": 1
    },
    "serial_number_format": {
      "prepend_if_missing": "91"
     }
  }
```

And for S10 it'd be

```
"validation": {
    "checksum": {
      "name": "s10"
    },
    "country_code": "exists"
  }
```

And then last final piece of the puzzle, which really outs me as a particularly pedantic variable namer…  all keys in the json files should be underscored (snake case) and not camel cased, no mixing. Which just means that our regex expressions will have to have named groups that are also underscored (like I had in my big json example up a few posts). I think the way the named group and the additional info section sync up is the way to go, while still being consistent with the rest.

I'm stoked about this though! This is great stuff, and we're gonna get written up in *Tracking Number Monthly* fo' sure.

**Comment by adgaudio** [comment]
I'm out for the next week until about Tuesday.  If you want to take a stab at it, go for it!  Otherwise, I'll continue on Tuesday.

**regex**   The regular expression named groups don't have underscores because java doesn't support underscores in named groups.  I could probably do something like replace all underscores in "?<strings_like_this_one>" but I suspect that will result in fairly messy code to process the regex and then link them to additional fields.  Do you mind if we leave the camel case named groups?  This would mean the keys in the "additional" section are also camel case.  Perhaps the solution is make the whole file camel case.

**misc.json**  :+1: Let's get rid of it :)

**validation** I didn't understand what you meant, but that example grouping makes sense to me.  Shouldn't "regex" go under validation too?

You know, I just overheard that you got promoted the Editor in Chief of *Tracking Number Monthly*!   Well done :smile:

**Comment by jkeen** [comment]
Got some stuff changed on this formatting and feel like I'm getting closer to being satisfied with the "additional" area. 

I'm starting to implement this into a branch of `tracking_number`, so some kinks should present themselves then.

**Comment by jkeen** [comment]
Oh yeah, I added some nested regex named groups for the UPS tracking number, similar to what you had in usps, but just a single string.

Did you have those usps regexes set up as arrays for a particular reason besides readability? I was initially trying to take each character group and wrap them in the `\b`'s for large text searching, which is why I had them split up into arrays in my initial example, but wondering if that's needed once I get this implementation shaken out.

**Comment by adgaudio** [comment]
This is looking great!  The regex as an array was only for readability.  It seems like your idea to wrap relevant sections in `\b` would be a great way to solve the whitespace problem!  I also like your direction on the additional fields, including getting rid of redundant regexes in s10, particularly since same logic applies in ups.  At first glance, "validation.additional.*" is a bit confusing, but I think I get it.

Looking forward to when you're finished working out additional integration details on your end - let me know when I can try re-integrating with java. 

Very cool, Jeff!  This is open source at its best :)

**Comment by jkeen** [comment]
Yeah! This is turning out to be really cool! 

I'm working on the ruby implementation and should have a better feel on where the kinks are by the end of this weekend.

**Comment by jkeen** [comment]
@adgaudio Do you know where you got the info for "FedEx Ground 96 (34)"? I'm looking for some further explanation of the character groups and can only find docs on the other Fedex Ground 96 barcode.

I've made some more changes that firm up some character group names so we're consistent in naming across different companies for the same info (e.g. `ServiceType` vs `ServiceIndentifier`), and I added another level of nesting for the tracking numbers so we have a place to put courier-related info.

**Comment by adgaudio** [comment]
Hi!  Sorry for the delay.  Fedex Ground (34) I think was the one I painstakingly pulled from the docs and also partly inferred or guessed parts from example tracking numbers I have that were not being recognized correctly

- Page 18 and 19 of http://images.fedex.com/us/solutions/ppe/FedEx_Ground_Label_Layout_Specification.pdf
- And this: http://www.fedex.com/us/gsn/barcode_guide.pdf

I suspect the serial number and check sum are correct, but other parts might be incorrect.  FedEx is a nightmare to say the least.

**Comment by jkeen** [comment]
Alright! I made some other changes you might want to look over and test out. I changed some regexes for USPS and some of the specs for the checksum. I'm gonna have to dig deeper and see if I missed anything in the USPS docs and then start working up our docs for these tracking numbers. I found some old USPS pdfs that have since been pulled from their website describing some of these number types.

Most importantly I modified all the regexes to allow optional spacing between the letters for detection. Just a single regex for each tracking number (which is an improvement from the way I had it in `tracking_number` needing two), but now it can match "9611020987654312345672" and "96 11 02  09  87 6 543 1234 56 72". 

The reason why it's important to allow spaces and not simply remove the spaces from the text to be searched (it turns out) came to light in the following test case that I had that kept failing

```ruby 
TrackingNumber.search("Lorem ipsum dolor sit amet, consectetur adipisicing elit,  
te, 1Z879E930346834440 nostru ommodo consequat. Duis aute 9611020987654312345672 
in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat 
cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.")
```
This was supposed to return two valid tracking numbers: a `UPS` one, `1Z879E930346834440`, and 
a `FedexGround` one `9611020987654312345672`. However, if you strip all the spaces out of the document, we can't detect for word boundaries, so *three* tracking numbers will be detected; the two previously mentioned, plus a valid DHL number *within* the UPS number.

Also, now with the spaces allowed I'm sure that the previous behavior in `tracking_number` will be maintained and consistent, and since some tracking numbers are formatted with spaces (USPS for example), this just feels like a pretty solid improvement. 

To detect/verify tracking numbers, the only changes I'm making internally to `tracking_number` are for searching to wrap the whole regex in `\b`, and for verifying start with `^` and end it with `$`. But I've got most of my tests passing on my feature branch of `tracking_number`, so things are progressing well. 

See if these changes work for you, more updates later.

**Comment by adgaudio** [comment]
Hi Jeff!

Sounds great!  I'll start working on integration next week on Tuesday.  It may take me a little bit to implement but I'll keep you updated :)

**Comment by adgaudio** [comment]
Hi Jeff!

I have some questions, but the one that I actually care about is your response the the "additional" section

- **`\\s*`**   shouldn't this just be `' *'`?  Do we want to regex match on new lines and tabs in numbers like "1 2\n3 4 5 6 7 ..."?
- **`\\s*`**  do you think all our valid number tests should include whitespace in them?  The ones we do have seem to be of the `xxxx xxxx` variety and not `x x x x x x x x`.
- How do you currently parse the matched regex with spaces? 
   I now do `" captured group  ".replaceAll(" ", "")` for every captured group which seems like a fair trade-off for the benefits we gained by including whitespace in regex, but maybe there's a better way...

- **additional** The additional fields are now an array with "regex_group_name" as the key.  The way we use these fields, ie for S10, is to search through the elements of that array looking for the `regex_group_name` that matches.  If, instead, `additional` was an object, the keys of which were the values of "regex_group_name", we would avoid doing a linear scan (O(1) vs O(n) lookup).   Was the reason we made the move from object to array to avoid having a camel case key?

Actually, looking into this further... I am finding it much more complicated code-wise to parse the json fields in java and link `validation.additional.exists` fields to the `additional.regex_group_name` fields.  Turns out java is just shy of hell to parse json with.  It would be simpler for me code-wise to revert to using an object, and then we also benefit by O(1) lookups:

    {validation: {additional: {exists: ['CapturedGroup']}},
      additional: {
          "CapturedGroup": {"AZ": {...}, "US": {...}}}
    }

If the only reason for this is camel case, would you be alright living with camel case keys for additional fields?  The change is also a fairly substantial reversion.  I can live with either, but this will take me a few days to sort out if we leave as is.  Let me know your thoughts.

**Comment by jkeen** [comment]
Hi Alex! I was actually in NYC last weekend but didn't have a chance to call you up and suggest meeting up. Turns out there is stuff to discuss!

1. `\\s*` Ohhh, that does new lines too. I think matching tabs is fine, maybe we can skip new lines. I just want to make sure we're liberal in the initial matching, as the validations should eliminate any false matches, and don't think we can rely on user input for *searching*. Does your port search swaths of unknown text and return all the tracking numbers within? Originally `tracking_number` was searching through emails and automatically tracking active numbers progress, and sometimes numbers were nested in a weird HTML table with god knows what between characters.

2. Another good point. Maybe we should format them in the ways that numbers are usually formatted. In my gem I've got tests that throw random spaces in a number to check detection, but this might be a good way to force the issue.

3. Yep, have to strip the spaces for the captured groups. I was trying to use a non-matching regex group to allow spaces between characters, but then have the named group not capture the spaces… but I could not get it to work correctly. That would be ideal if we could figure that out.

4. Additional: I hear you on this. I wanted it to be a simple lookup too, but the problem I was trying to address (and still haven't successfully arrived on a solution) is how to handle when one character group (e.g. `CountryCode`) is responsible for associating multiple pieces of information. By making additional an array and having `regex_group_name` as a key, it allows multiple entries to derive data from the same captured group. Same with `matches` and `matches_regex`. I wanted to make it as explicit and obvious as possible from just reading the json on how to work the thing, rather than having to document "when the value of `additional` is an object do X, when it's an array, do Y".

I don't know java, but the way my implementation is working is I just read in the file once and create the classes, so the O(n) O(1) thing was just a little extra hassle, and not a performance issue. But ruby is pretty open and metaprogramming is a thing there, and Java is a different beast I'm sure. But could you potentially do the same thing, eliminating O(n)/O(1) worries on runtime, and read in the json, make those things hashes for easy lookups and then continue on your way? I'll keep thinking on this.

**Comment by adgaudio** [comment]
Next time you are in town, I'd love to get together!  

1. **`\\s*`** Since you mentioned scraping HTML that has newlines, lets keep newlines in there and use `\\s*` as is.  I realized users can simply format the tracking number strings ahead of time if they don't want to consider newlines.  For instance, one could do `map(parse_tracking_number, string.split('\n'))`.
2. I'll add spaces on some of the test numbers to explicitly account for it.
3. ... we're both fine with stripping whitespaces.  :)
4. **additional** - Got it working.  I guess I don't understand how `CountryCode` is responsible for associating multiple pieces of information, but maybe you can help me understand whether I'm doing this correctly:  When I parse the json files, I convert the arrays to hash tables.  My assumption is that the key, `additional[].name`, uniquely identifies each element in the `additional` array, and secondly, either the key `additional[].lookup[].matches` xor the key `additional[].lookup[].matches_regex`, uniquely identifies each element in the `lookup` array.  If this assumption is valid, what do the arrays help us do?  I'm happy to keep what we have currently, but I would like to understand why it's in this form as it seems like things are unnecessarily complicated.

---

5. USPS91 and FedEx SmartPost now share the same tracking number.  Is that the behavior we want?  I guess it probably is, though now my tests are failing because SmartPost tracking numbers parse as USPS91.  How did you handle this?

**Comment by jkeen** [comment]
Sounds great re: meeting up, and great that you got it working the way it is. 

4. Additional and multiple pieces of info: 

I was trying to figure out a way to have a consistent way of bubbling up additional data. For instance in regards to `ServiceType`, one service code generally relates to a single string describing that class of service—like "UPS Next Day", or "Letter Post Registered". So in my implementation was automatically creating a method `service_type` that would pull a `name` property matched lookup values in the `additional` data section so I could do:

```ruby
TrackingNumber.new('1325125125125215').service_type 
#=> "Next Day Air"  
``` 
but where it got tricky was the cases where there's a name + a description from that one piece of data, or in the case of country code, where there's the name of the country, the name of the courier, and a url to the courier from that two character country code. And I hadn't yet figured out if I was just going to return a hash of values, 
```ruby
TrackingNumber.new('1325125125125215').service_type 
#=> {name: "Letter Post Registered" , description: "Prepaid first-class mail that is recorded by the post office before being sent and at each point along its route to safeguard against loss, theft, or damage"}
``` 

or implement a different method for each discrete piece of data:
```ruby
t= TrackingNumber.new('1325125125125215')
t.service_type
#=> "Letter Post Registered"  
t.service_type_description
#=> "Prepaid first-class mail that is recorded by the post office before being sent and at each point along its route to safeguard against loss, theft, or damage"
``` 

And with country code indicating the Country, the name of the Courier, and the url of the courier, I was debating if the `additional` format in this library should only allow one value per match to be returned, somewhat ensuring a consistent interface between implementations? Which is why I wanted the array with the value of a the key determining a match group to leave that option open. 

```json
[
  {         
    "name": "Courier",
    "regex_group_name": "CountryCode",
    "lookup": {
      "UA": "Ukrposhta",
      "AE": "Emirates Post",
      "US": "United States Postal Service"
    }
  },
  {         
    "name": "Country",
    "regex_group_name": "CountryCode",
    "lookup": {
      "UA": "Ukraine",
      "AE": "United Arab Emirates",
      "US": "United States of America"
    }
  },
  {
    "name": "Courier URL",
    "regex_group_name": "CountryCode",
    "lookup": {
      "UA": "http://services.ukrposhta.com/postindex_new/default.aspx?lang=en",
      "AE": "http://www.emiratespost.com/content/english/index.jsp",
      "US": "http://zip4.usps.com/zip4/welcome.jsp"  
    }
  }
]

```

Hmm, I wasn't sure of that direction but after typing it out… maybe that'd be good?

Also, the sender/deliverer problem [described here](https://github.com/jkeen/tracking_number/issues/27) might require `ServiceType` indicating the type of service as well as the shipper. Have you found anything more about that issue in your research? That was part of the reason for the the SmartPost addition which unintentionally caused an overlap with USPS91. I'll have to get my head back into this again.

Sorry for the wordy update again, hope that makes sense

**Comment by adgaudio** [comment]
I think I understand what you just said.  To summarize:

1. what should `s10_tracking_number.service_type` return, a dictionary or a string? 
2. How should we format the additional field in json files?

---

1. I think it should return a dictionary like you presented:

    > TrackingNumber.new('1325125125125215').service_type
    > #=> {name: "Letter Post Registered",
    >      description: "Prepaid first-class mail that is recorded by the post office before being sent and at each point along its route to safeguard against loss, theft, or damage"}

and your first example would be like this

```
TrackingNumber.new('1325125125125215').service_type 
#=> {name: "Next Day Air"}
```

The idea of the `additional` field, in my mind, is to summarize all the extra information we have about some courier in one place.  Applied to a tracking number, the additional field serves to a) validate the tracking number (ie `validation[].exists`) and b) provide the relevant extra data associated to that tracking number.  Therefore, it seems natural that the TrackingNumber.parse("blah") internally represent a dictionary, and the fields on that object are the keys of additional we care about.  

If the JSON file looked like this:
```
{
  "ServiceType": {
    "E[A-Z]": {name: "EMS", description: "blah..."},
     ...
  },
  "CountryCode": {
    "AF": {country: "Afghanistan", courier: "Afghan Post", ...},
    ...
  }
}
```

Then `TrackingNumber.new('1325125125125215').service_type` simply returns the matching dictionary.  Likewise, `TrackingNumber.new('1325125125125215').country_code` also returns the matching dictionary.


2. I think your latter example of the json file seems more like the format that your code might parse the json file into.  Since that form would be harder to maintain, I prefer what we currently have.  


If you don't agree, I'm totally game to keep what we currently have, since both of us have a working implementation on it :)

**Comment by jkeen** [comment]
Yeah, I got to thinking more, and it might be too much to make this dictate how the client libraries implement their interfaces. For the ruby implementation I'm going to make methods like I was saying, so that `.service_type` always returns a string if it's present. But different languages have different styles, and returning a dictionary might be what's best for your deal, so I'm cool with leaving it as is.

After I finish the ruby implementation I want to make a javascript implementation and a browser-based tracking number decoder, so that'll give me a better sense of how this feels in other languages.

But anyway, I think we can close out this PR, consider this a good starting point, and continue in the github issues with further changes. Agree?

**Comment by adgaudio** [comment]
Ok this sounds good to me!  :)

When you implement the javascript version, if you find some configuration that you like better, I'll change the java library to match the format you end up with.  Also, I was thinking a very simple Python version will be very helpful for some analysis I'm doing, so I might get around to that some time, too!

Also, I'm glad you figured out this whitespace and searching issue because I have a bunch of 2d barcodes that get scanned, and we need to search through the text to find the tracking number.

enthusiastic :+1: !
